blazar/enforcement/filters/max_reservation_length_filter.py

```diff
@@ -0,0 +1,99 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import datetime, timedelta
+
+from blazar import exceptions
+from blazar.i18n import _
+from blazar.enforcement.filters import base_filter
+
+from oslo_config import cfg
+from oslo_log import log as logging
+
+
+DEFAULT_MAX_RESERVATION_LENGTH = 0
+DEFAULT_RESERVATION_EXTENTSION_WINDOW = -1
+
+
+LOG = logging.getLogger(__name__)
+
+
+class MaxReservationLengthException(exceptions.BlazarException):
+    code = 400
+    msg_fmt = _('Lease length of %(lease_length)s seconds be less than or '
+                'equal the maximum lease length of %(max_length)s seconds.')
+
+
+class MaxReservationUpdateWindowException(exceptions.BlazarException):
+    code = 400
+    msg_fmt = _('Lease can only be extended within %(extension_window)s '
+                'seconds of the leases current end time.')
+
+
+class MaxReservationLengthFilter(base_filter.BaseFilter):
+
+    enforcement_opts = [
+        cfg.IntOpt(
+            'max_reservation_length',
+            default=DEFAULT_MAX_RESERVATION_LENGTH,
+            help='Maximum lease duration in seconds. If this is set to -1, '
+                 'there is not limit. For active leases being updated, '
+                 'the limit applies between now and the new end date.'),
+        cfg.IntOpt(
+            'reservation_extension_window',
+            default=DEFAULT_RESERVATION_EXTENTSION_WINDOW,
+            help='Gives users a window towards the end of a reservation to '
+                 'extend their lease again for the max lease length. A '
+                 'value of -1 will not allow users to extend leases beyond '
+                 'the maximum lease length')
+    ]
+
+    def __init__(self, conf=None):
+        super(MaxReservationLengthFilter, self).__init__(conf=conf)
+
+    def check_for_length_violation(self, start_date, end_date):
+        lease_length = (end_date - start_date).total_seconds()
+
+        if lease_length > self.conf.enforcement.max_reservation_length:
+            raise MaxReservationLengthException(
+                lease_length=lease_length,
+                max_length=self.conf.enforcement.max_reservation_length)
+
+    def check_create(self, context, lease_values):
+        start_date = lease_values['start_date']
+        end_date = lease_values['end_date']
+
+        self.check_for_length_violation(start_date, end_date)
+
+    def check_update(self, context, current_lease_values, new_lease_values):
+        start_date = current_lease_values['start_date']
+        end_date = new_lease_values['end_date']
+
+        if self.conf.enforcement.reservation_extension_window:
+            min_window = current_lease_values['end_date'] - timedelta(
+                seconds=self.conf.enforcement.reservation_extension_window)
+            update_at = datetime.utcnow()
```

Comment:
How does Blazar handle timestamps elsewhere? We should follow whatever it does, maybe it is this mechanism. The important part is that all pieces agree. There are subtle differences in utcnow vs using pytz.UTC for example.